File: lib/features/planned_changes/domain/models/planned_change.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'planned_change.freezed.dart';
part 'planned_change.g.dart';

@freezed
abstract class PlannedChange with _$PlannedChange {
  const factory PlannedChange({
    required String id,
    required String parentId, // Income.id or Expense.id
    required String parentType, // 'income' or 'expense'
    required double newAmount, // new amount from effectiveDate
    required DateTime effectiveDate,
    @Default(false) bool isGross, // for income: is newAmount gross?
    String? note,
    @Default(false) bool isDeleted,
    required DateTime createdAt,
  }) = _PlannedChange;

  factory PlannedChange.fromJson(Map<String, dynamic> json) =>
      _$PlannedChangeFromJson(json);
}

File: SavvyApp/Packages/Core/SavvyFoundation/Sources/SavvyFoundation/Models/PlannedChange.swift
import Foundation

public struct PlannedChange: FirestoreEntity {
    public let id: String
    public var parentId: String
    public var parentType: String
    public var newAmount: Decimal
    public var effectiveDate: Date
    public var isGross: Bool
    public var note: String?
    public var isDeleted: Bool
    public let createdAt: Date

    public init(
        id: String = UUID().uuidString,
        parentId: String,
        parentType: String,
        newAmount: Decimal,
        effectiveDate: Date,
        isGross: Bool = false,
        note: String? = nil,
        isDeleted: Bool = false,
        createdAt: Date = Date()
    ) {
        self.id = id
        self.parentId = parentId
        self.parentType = parentType
        self.newAmount = newAmount
        self.effectiveDate = effectiveDate
        self.isGross = isGross
        self.note = note
        self.isDeleted = isDeleted
        self.createdAt = createdAt
    }
}
